File: proxy-android/proxy/src/main/java/com/zeticai/melangelm/upstream/OpenAIUpstreamClient.kt
package com.zeticai.melangelm.upstream

import com.zeticai.melangelm.model.ChatChoice
import com.zeticai.melangelm.model.ChatMessage
import com.zeticai.melangelm.model.ChatRequest
import com.zeticai.melangelm.model.ChatResponse
import com.zeticai.melangelm.model.TokenUsage
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * OpenAI-compatible upstream client.
 * Works with OpenAI, Azure OpenAI, Anthropic (via proxy), and any compatible endpoint.
 */
class OpenAIUpstreamClient(
    private val baseUrl: String,
    private val apiKey: String,
    private val defaultModel: String? = null,
    timeoutSeconds: Long = 60
) : UpstreamClient {

    private val http = OkHttpClient.Builder()
        .connectTimeout(timeoutSeconds, TimeUnit.SECONDS)
        .readTimeout(timeoutSeconds, TimeUnit.SECONDS)
        .writeTimeout(timeoutSeconds, TimeUnit.SECONDS)
        .build()

    private val JSON = "application/json; charset=utf-8".toMediaType()

    override suspend fun send(request: ChatRequest): ChatResponse {
        val body = buildRequestJson(request)
        val httpRequest = Request.Builder()
            .url("${baseUrl.trimEnd('/')}/v1/chat/completions")
            .addHeader("Authorization", "Bearer $apiKey")
            .addHeader("Content-Type", "application/json")
            .post(body.toString().toRequestBody(JSON))
            .build()

        val httpResponse = http.newCall(httpRequest).execute()
        val responseBody = httpResponse.body?.string()
            ?: throw RuntimeException("Empty response from upstream (HTTP ${httpResponse.code})")

        if (!httpResponse.isSuccessful) {
            throw RuntimeException("Upstream HTTP ${httpResponse.code}: $responseBody")
        }

        return parseResponse(responseBody)
    }

    private fun buildRequestJson(request: ChatRequest): JSONObject {
        val messages = JSONArray()
        request.messages.forEach { msg ->
            messages.put(JSONObject().apply {
                put("role", msg.role)
                put("content", msg.content)
            })
        }
        return JSONObject().apply {
            put("model", defaultModel ?: request.model)
            put("messages", messages)
            request.temperature?.let { put("temperature", it) }
            request.maxTokens?.let { put("max_tokens", it) }
            if (request.stream) put("stream", true)
        }
    }

    private fun parseResponse(json: String): ChatResponse {
        val root = JSONObject(json)
        val choicesArr = root.getJSONArray("choices")
        val choices = (0 until choicesArr.length()).map { i ->
            val c = choicesArr.getJSONObject(i)
            val msg = c.getJSONObject("message")
            ChatChoice(
                index = c.optInt("index", i),
                message = ChatMessage(
                    role = msg.getString("role"),
                    content = msg.getString("content")
                ),
                finishReason = c.optString("finish_reason")
            )
        }
        val usage = root.optJSONObject("usage")?.let {
            TokenUsage(
                promptTokens = it.optInt("prompt_tokens"),
                completionTokens = it.optInt("completion_tokens"),
                totalTokens = it.optInt("total_tokens")
            )
        }
        return ChatResponse(
            id = root.optString("id", ""),
            model = root.optString("model", ""),
            choices = choices,
            usage = usage
        )
    }
}

File: proxy-ios/Sources/MelangeLmProxy/Upstream/OpenAIUpstreamClient.swift
//
//  OpenAIUpstreamClient.swift
//  MelangeLmProxy
//
//  OpenAI-compatible upstream client using URLSession.
//  Works with OpenAI, Azure OpenAI, Anthropic (via proxy), and any /v1/chat/completions endpoint.
//

import Foundation

public final class OpenAIUpstreamClient: UpstreamClient {
    private let baseURL: URL
    private let apiKey: String
    private let defaultModel: String?
    private let session: URLSession

    public init(
        baseURL: String,
        apiKey: String,
        defaultModel: String? = nil,
        timeoutSeconds: Double = 60
    ) {
        guard let url = URL(string: baseURL) else {
            fatalError("MelangeLmProxy: invalid upstream baseURL: \(baseURL)")
        }
        self.baseURL = url
        self.apiKey = apiKey
        self.defaultModel = defaultModel
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = timeoutSeconds
        config.timeoutIntervalForResource = timeoutSeconds
        self.session = URLSession(configuration: config)
    }

    public func send(_ chatRequest: ChatRequest) async throws -> ChatResponse {
        let endpoint = baseURL
            .appendingPathComponent("v1")
            .appendingPathComponent("chat")
            .appendingPathComponent("completions")

        var urlRequest = URLRequest(url: endpoint)
        urlRequest.httpMethod = "POST"
        urlRequest.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        urlRequest.httpBody = try buildBody(chatRequest)

        let (data, response) = try await session.data(for: urlRequest)

        if let http = response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
            let body = String(data: data, encoding: .utf8) ?? "(no body)"
            throw UpstreamError.httpError(statusCode: http.statusCode, body: body)
        }

        return try parseResponse(data)
    }

    // MARK: - Encoding

    private func buildBody(_ request: ChatRequest) throws -> Data {
        var json: [String: Any] = [
            "model": defaultModel ?? request.model,
            "messages": request.messages.map { ["role": $0.role, "content": $0.content] }
        ]
        if let temperature = request.temperature { json["temperature"] = temperature }
        if let maxTokens = request.maxTokens { json["max_tokens"] = maxTokens }
        if request.stream { json["stream"] = true }
        return try JSONSerialization.data(withJSONObject: json)
    }

    // MARK: - Decoding

    private func parseResponse(_ data: Data) throws -> ChatResponse {
        guard let root = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw UpstreamError.malformedResponse("Could not parse root object")
        }
        guard let choicesArr = root["choices"] as? [[String: Any]] else {
            throw UpstreamError.malformedResponse("Missing 'choices'")
        }

        let choices: [ChatChoice] = try choicesArr.enumerated().map { (i, c) in
            guard let msgDict = c["message"] as? [String: Any],
                  let role = msgDict["role"] as? String,
                  let content = msgDict["content"] as? String else {
                throw UpstreamError.malformedResponse("Malformed choice at index \(i)")
            }
            return ChatChoice(
                index: c["index"] as? Int ?? i,
                message: ChatMessage(role: role, content: content),
                finishReason: c["finish_reason"] as? String
            )
        }

        var usage: TokenUsage?
        if let u = root["usage"] as? [String: Any] {
            usage = TokenUsage(
                promptTokens: u["prompt_tokens"] as? Int ?? 0,
                completionTokens: u["completion_tokens"] as? Int ?? 0,
                totalTokens: u["total_tokens"] as? Int ?? 0
            )
        }

        return ChatResponse(
            id: root["id"] as? String ?? "",
            model: root["model"] as? String ?? "",
            choices: choices,
            usage: usage
        )
    }
}

public enum UpstreamError: Error, LocalizedError {
    case httpError(statusCode: Int, body: String)
    case malformedResponse(String)

    public var errorDescription: String? {
        switch self {
        case .httpError(let code, let body): return "HTTP \(code): \(body)"
        case .malformedResponse(let detail): return "Malformed response: \(detail)"
        }
    }
}
